File: FindEquilibria.m
%% Parameters to tune: rates of snake depredation of eagle (X) and deer (Y)
X = 0.0;
Y = 0.0;
%________________________________________________________________________
%% Not to tune: these are our system dynamics 

% Parameters of the separated predator-prey systems
b1 = 4;
b2 = -1;
b3 = 2;
b4 = -1;
a12 = 0.5;
a21 = -1;
a34 = 0.5;
a43 = -1;

% Parameters that we tune from the snake entering the second system
a23 = -X; % Snake eats deer
a32 = X/2; % Deer is eaten by snake
a24 = -Y; % Snake eats eagle
a42 = Y/2; % Eagle is eaten by snake

% Parameters we don't consider
a11 = 0;
a22 = 0;
a33 = 0;
a44 = 0;
a13 = 0;
a14 = 0;
a41 = 0;
a31 = 0;

%% Find the equilibrium
b = [b1; b2; b3; b4];

A = [a11 a12 a13 a14;
     a21 a22 a23 a24;
     a31 a32 a33 a34;
     a41 a42 a43 a44];
 
eq = inv(A) * b

%Loop for studying the effect of changing eigenvalues

n_iter = 10;

eig_tab = cell(1,n_iter);
div_tab = cell(1,n_iter);

figure()

for n = 1:n_iter
    
    X = X + 0.05;
    
    Y = Y + 0.025;
    
    a23 = -X; % Snake eats deer
    a32 = X/2; % Deer is eaten by snake
    a24 = -Y; % Snake eats eagle
    a42 = Y/2; % Eagle is eaten by snake

    
    b = [b1; b2; b3; b4];

    A = [a11 a12 a13 a14;
     a21 a22 a23 a24;
     a31 a32 a33 a34;
     a41 a42 a43 a44];
    
   eq = inv(A) * b;
    
   [check_deval, check_eigs, div] = linearise(eq,A,b);
   
   eig_tab{n} = check_eigs;
   div_tab{n} = div;
   
   %Plot the eigenvalues
   hold on
   string = sprintf('X= %G Y =%G', X, Y);
   plot(eig_tab{n}, 'o','DisplayName', [string]);
   xlabel('Real')
   ylabel('Imaginary')
   legend('show')
  
end

xline(0)
yline(0)

%Check for global stability...to be implemented


%Extension to equilibria investigation 

%Plug in X and Y values and investigate whether equilibria is stable? 

%% Linearise about equlibria. compute eigenvalues and divergence

function[dd_eval, Eigens, div] = linearise(equilibria, A, b)

syms x y z w

dx = x * (b(1) - A(1,1)*x - A(1,2)*y - A(1,3)*z - A(1,4)*w);
dy = y * (b(2) - A(2,1)*x- A(2,2)*y - A(2,3)*z - A(2,4)*w);
dz = z * (b(3) - A(3,1)*x - A(3,2)*y - A(3,3)*z - A(3,4)*w);
dw = w * (b(4) - A(4,1)*x - A(4,2)*y - A(4,3)*z - A(4,4)*w);

%Matrix of partials:

dd = [diff(dx,x) diff(dx,y) diff(dx,z) diff(dx,w);
    diff(dy,x) diff(dy,y) diff(dy,z) diff(dy,w);
    diff(dz,x) diff(dz,y) diff(dz,z) diff(dz,w)
    diff(dw,x) diff(dw,y) diff(dw,z) diff(dw,w);];

%Evaluate the matrix at the equilibrium points: 

dd_eval = ones(4,4);

x0 = equilibria(1);
y0 = equilibria(2);
z0 = equilibria(3);
w0 = equilibria(4);


for i = 1:4
    for j = 1:4
        dd_eval(i,j) = subs(dd(i,j), {x, y, w, z}, {x0, y0, z0, w0});      
    end
end

%Compute the matrix eigenvalues:
Eigens = eigs(dd_eval);

%Compute the divergence
div = diff(dx,x) + diff(dy,y) +  diff(dz,z) + diff(dw,w);
end 


%% Computing the divergence in order to check for a non-existence of the limit cycle

 

File: src/FindEquilibria.py
from ConfigLoader import ConfigLoader
import numpy as np
from numpy import linalg as LA
from Plotter import Plotter

class FindEquilibria:
    '''
    this class implements the algorithms to find the equilibria of the system
    '''
    def __init__(self):
        config = ConfigLoader()
        self.config = config.params
        self.a, self.b = self.setParameters()
        self.plotter = Plotter()
        self.k, self.sigma = self.setPositiveControlParameters()

    def setParameters(self):
        a = [[self.config['a11'], self.config['a12'], self.config['a13'], self.config['a14']], \
            [self.config['a21'], self.config['a22'], self.config['a23'], self.config['a24']], \
            [self.config['a31'], self.config['a32'], self.config['a33'], self.config['a34']], \
            [self.config['a41'], self.config['a42'], self.config['a43'], self.config['a44']]]
        b = [self.config['b1'], self.config['b2'], self.config['b3'], self.config['b4']]

        if 'X' in self.config and 'Y' in self.config:
            X = self.config['X']
            Y = self.config['Y']
            a[1][2] = -X
            a[2][1] = X/2
            a[1][3] = -Y
            a[3][1] = Y/2
        return a, b
    
    def setPositiveControlParameters(self):
        k = [self.config['k1'], self.config['k2'], self.config['k3'], self.config['k4']]
        sigma = self.config['sigma']
        return k, sigma

    def setParametersAgain(self):
        self.a, self.b = self.setParameters()

    def findEquilibria(self, printEq=True):
        eq = np.matmul(np.linalg.inv(np.array(self.a)), self.b).tolist()
        if self.config['printEquilibria']:
            print('The equilibria of the system are:\n')
            if printEq:
                for el in eq:
                    print(str(el) + '\n')
        return eq

    def linearise(self, equilibrium, considerControl:bool=False, *args):
        # linearised dynamics 
        b = self.b
        a = self.a
        k = self.k
        x0 = equilibrium[0]
        y0 = equilibrium[1]
        z0 = equilibrium[2]
        w0 = equilibrium[3]
        for el in args:
            u = el
            break
        if self.config['useControl'] and considerControl and self.config['usePositiveControl']:
            dyn_lin = [[b[0] - 2*a[0][0]*x0 - a[0][1]*y0 - a[0][2]*z0 - a[0][3]*w0 + k[0]*u, -a[0][1]*x0, -a[0][2]*x0, -a[0][3]*x0], \
            [-a[1][0]*y0, b[1] - a[1][0]*x0 - 2*a[1][1]*y0 - a[1][2]*z0 - a[1][3]*w0 + k[1]*u, -a[1][2]*y0, -a[1][3]*y0], \
            [-a[2][0]*z0, -a[2][1]*z0, b[2] - a[2][0]*x0 - a[2][1]*y0 - 2*a[2][2]*z0 - a[2][3]*w0 + k[2]*u, -a[2][3]*z0], \
            [-a[3][0]*w0, -a[3][1]*w0, -a[3][2]*w0, b[3] - a[3][0]*x0 - a[3][1]*y0 - a[3][2]*z0 - 2*a[3][3]*w0 + k[3]*u]]
        else:
            dyn_lin = [[b[0] - 2*a[0][0]*x0 - a[0][1]*y0 - a[0][2]*z0 - a[0][3]*w0, -a[0][1]*x0, -a[0][2]*x0, -a[0][3]*x0], \
                [-a[1][0]*y0, b[1] - a[1][0]*x0 - 2*a[1][1]*y0 - a[1][2]*z0 - a[1][3]*w0, -a[1][2]*y0, -a[1][3]*y0], \
                [-a[2][0]*z0, -a[2][1]*z0, b[2] - a[2][0]*x0 - a[2][1]*y0 - 2*a[2][2]*z0 - a[2][3]*w0, -a[2][3]*z0], \
                [-a[3][0]*w0, -a[3][1]*w0, -a[3][2]*w0, b[3] - a[3][0]*x0 - a[3][1]*y0 - a[3][2]*z0 - 2*a[3][3]*w0]]
        if self.config['printEquilibria']:
            print(equilibrium)
            print('---')
            print(dyn_lin)
        return dyn_lin
    
    def getEigens(self):
        eq = self.findEquilibria()
        dyn_lin = self.linearise(eq)
        eigenvalues, _ = LA.eig(dyn_lin)
        if self.config['plotEigenvalues'] and not self.config['plotStabilityXY']:
            self.plotter.plotEigen(eigenvalues)
        return eigenvalues
